tools/Customization/DevHome.Customization/Views/FileExplorerPage.xaml.cs

```diff
@@ -1,23 +1,83 @@
-// Copyright (c) Microsoft Corporation.
-// Licensed under the MIT License.
-
-using DevHome.Common.Extensions;
-using DevHome.Customization.ViewModels;
-using Microsoft.UI.Xaml;
-using Microsoft.UI.Xaml.Controls;
-
-namespace DevHome.Customization.Views;
-
-public sealed partial class FileExplorerPage : Page
-{
-    public FileExplorerViewModel ViewModel
-    {
-        get;
-    }
-
-    public FileExplorerPage()
-    {
-        ViewModel = Application.Current.GetService<FileExplorerViewModel>();
-        this.InitializeComponent();
-    }
-}
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using DevHome.Common.Extensions;
+using DevHome.Common.Services;
+using DevHome.Customization.ViewModels;
+using DevHome.FileExplorerSourceControlIntegration.Services;
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+
+namespace DevHome.Customization.Views;
+
+public sealed partial class FileExplorerPage : Page
+{
+    public FileExplorerViewModel ViewModel
+    {
+        get;
+    }
+
+    private string? RootPath
+    {
+        get; set;
+    }
+
+    private static readonly System.Buffers.SearchValues<char> InvalidChars = System.Buffers.SearchValues.Create("<>*?|.");
+
+    public FileExplorerPage()
+    {
+        ViewModel = Application.Current.GetService<FileExplorerViewModel>();
+        this.InitializeComponent();
+        var experimentationService = Application.Current.GetService<IExperimentationService>();
+        var isEnabled = experimentationService.IsFeatureEnabled("FileExplorerSourceControlIntegration");
+        if (isEnabled)
+        {
+            TrackRepository.Visibility = Visibility.Visible;
+            DisplayTrackRepository.Visibility = Visibility.Visible;
+        }
+    }
+
+    public void AddButton_Click(object sender, RoutedEventArgs e)
+    {
+        RootPath = RootPathTextBox.Text;
+
+        var isValidated = ValidateRepositoryPath(RootPath);
+        if (isValidated)
+        {
+            // TO DO: Determine if the extension GUID should be stored instead of name as it is more identifiable and/or determine any other unique property to use for
+            // mapping here
+            ViewModel.AddRepositoryPath("git", RootPath);
+        }
+
+        ViewModel.RefreshTrackedRepositories();
+    }
+
+    public bool ValidateRepositoryPath(string rootPath)
+    {
+        if (!Path.IsPathFullyQualified(rootPath))
+        {
+            // Error: Path is not fully qualified or maybe a UNC path.
+            RootPathErrorBar.IsOpen = true;

```

Comment:
NIT: Change the error message based on what is being checked.  Give the user the exact reason why the path isn't valid.

Reply:
> NIT: Change the error message based on what is being checked. Give the user the exact reason why the path isn't valid.

+1 to this. We may even consider having some different error/exception types for path checking.

Once interesting example is that repositories mounted on WSL look sorta like UNC paths, but we should try to detect that case separate from UNC paths. Also with WSL repos, their file ownership looks off from Windows' point of view, so Git will raise specific errors about those cases. Rather than outright failing, we should offer a suggestion to the user to help them enable Git to work with those WSL repos, aka `git config --global --add safe.directory <path to repo>`

Reply:
Sounds great! Currently, the UI in place is temporary. I will incorporate this feedback and gain more clarity around the design for surfacing user facing error messages in the expected UX.